transaction/std/src/memo_builder/gift_code_memo_builder.rs

```diff
@@ -0,0 +1,181 @@
+            wrote_change_memo: false,
+        }
+    }
+    /// A note in the form of null terminated utf-8 bytes can be applied to
+    /// gift code funding or sender memos. These notes end up in the change
+    /// outputs and indicate what the memo was for. Set this memo here. This
+    /// method will enforce the 60 and 64 byte length for Funding Memos and
+    /// Sender Memos respectively. If the memo type is a cancellation memo
+    /// and a note is set, any value set here will be ignored as the
+    /// cancellation memo does not include a note.
+    pub fn set_note(&mut self, note: &'static str) -> Result<(), MemoError> {
+        let max_note_length = if let GiftCodeMemoType::Funding = self.gift_code_memo_type {
+            60
+        } else {
+            64
+        };
+        if note.len() > max_note_length {
+            return Err(MemoError::BadLength(note.len()));
+        }
+        self.note = note;
+        Ok(())
+    }
+    /// Clear the gift code note
+    pub fn clear_note(&mut self) {
+        self.note = "";
+    }
+    /// Set tx_out_public_key to be written to a sender memo. If the type is not
+    /// a sender memo the key isn't written.
+    pub fn set_gift_code_tx_out_public_key(&mut self, tx_out_public_key: RistrettoPublic) {
+        if let GiftCodeMemoType::Sender = self.gift_code_memo_type {
+            self.gift_code_tx_out_public_key = Some(tx_out_public_key);
+        }
+    }
+    /// Clear the gift code tx_out_public_key
+    pub fn clear_gift_code_tx_out_public_key(&mut self) {
+        self.gift_code_tx_out_public_key = None;
+    }
+    /// Set global index for the TxOut
+    pub fn set_gift_code_tx_out_index(&mut self, tx_out_global_index: u64) {
+        self.gift_code_tx_out_global_index = Some(tx_out_global_index)
+    }
+    /// Clear global index for gift code TxOut
+    pub fn clear_gift_code_tx_out_index(&mut self) {
+        self.gift_code_tx_out_global_index = None;
+    }
+    /// Enable gift code change memos
+    pub fn enable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = true;
+    }
+    /// Disable gift code change memos
+    pub fn disable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = false;
+    }
+}
+
+impl MemoBuilder for GiftCodeMemoBuilder {
+    /// Set the fee
+    fn set_fee(&mut self, _fee: Amount) -> Result<(), NewMemoError> {
+        Ok(())
+    }
+
+    /// Gift code destination memos are not allowed - RTH transaction builders
+    /// should be used alongside this builder to create recoverable
+    /// transaction history if desired
+    fn make_memo_for_output(
+        &mut self,
+        _amount: Amount,
+        _recipient: &PublicAddress,
+        _memo_context: MemoContext,
+    ) -> Result<MemoPayload, NewMemoError> {
+        Err(NewMemoError::GiftCodeDesintationMemoNotAllowed)
+    }
+
+    /// Build a memo for a gift code change output (to ourselves)
+    fn make_memo_for_change_output(
```

Comment:
I appreciate the value of breaking up changes, but I'm not seeing anything exercising/testing this behavior.  Is there  a plan to help ensure correct behavior going forward?

Reply:
Yeah there will be - but I want to get feedback on the design first to ensure its sanity checked.

Reply:
Tests now incoming!